backend/coreapp/views.py

```diff
@@ -201,3 +216,91 @@ def fork(request, slug):
+
+@api_view(["GET", "POST"])
+def user_current(request, slug=None):
+    """
+    Get the logged-in user, or sign in with GitHub
+    """
+
+    profile = Profile.objects.filter(id=request.session.get("profile", None)).first()
+
+    if not profile:
+        profile = Profile()
+        profile.save()
+        request.session["profile"] = profile.id
+
+    # sign in
+    if request.method == "POST":
+        if not settings.GITHUB_CLIENT_ID or not settings.GITHUB_CLIENT_SECRET:
+            return Response({
+                "error": "GitHub sign-in not configured"
+            }, status=status.HTTP_501_NOT_IMPLEMENTED)
+
+        required_params = ["code"]
+
+        for param in required_params:
+            if param not in request.data:
+                return Response({"error": f"Missing parameter: {param}"}, status=status.HTTP_400_BAD_REQUEST)
+
+        logging.debug("Attempting GitHub oauth login")
+
+        response = requests.post(
+            "https://github.com/login/oauth/access_token",
+            json={
+                "client_id": settings.GITHUB_CLIENT_ID,
+                "client_secret": settings.GITHUB_CLIENT_SECRET,
+                "code": request.data["code"],
+            },
+            headers={ 'Accept': 'application/json' },
+        ).json()
+
+        error = response.get("error")
+        if error == None:
+            access_token = response["access_token"]
+
+            profile.github_access_token = access_token
+
+            try:
+                assert profile.load_fields_from_github(always=True)
+                logging.debug("Connected existing profile to new GitHub user")
+            except GitHubUserChangeException:
+                # The token was for a different user than the one associated with the current profile,
+                # so make a new profile for this one.
+                profile = Profile()
+                profile.github_access_token = access_token
+                assert profile.load_fields_from_github(always=True)
+                logging.debug("Connected new profile to new GitHub user")
+            except GitHubUserHasExistingProfileException as e:
+                profile = e.profile
+
+                # This isn't strictly necessary, but we might as well use the renewed access token.
+                profile.github_access_token = access_token
+                assert profile.load_fields_from_github(always=True)
+                logging.debug("Swapped to existing profile for existing GitHub user")
+
+            profile.save()
+            request.session["profile"] = profile.id
+        elif error == "bad_verification_code":
+            return Response({
+                "error": "Invalid or expired GitHub OAuth verification code",
+            }, status=status.HTTP_401_UNAUTHORIZED)
+        else:
+            raise Exception(f"Unknown GitHub login error: {error} - {response['error_description']}")
+
+    else:
```

Comment:
Since this is far from the original if and it's not completely obvious that this has to be a GET, I think it's worth doing elif GET here

Reply:
:+1:

Reply:
Split GET/POST handlers into separate functions? I've wanted to do that for the `/api/scratch` route.

Reply:
Yeah that makes sense actually, particularly for /api/scratch

Reply:
There's a Django view class specifically for this, I think, so I'll use that